coins-on-the-clock/rust/assembly.s

```diff
@@ -336,166 +334,162 @@ example::_get_valid_sequences:
         mov     dl, al
         shl     rdx, 3
         lea     rsi, [rsi + 2*rsi]
-        mov     rax, qword ptr [rsp + 24]
-        mov     rdi, qword ptr [rax]
-        mov     qword ptr [rsp], rsi
-        mov     qword ptr [rsp + 8], rdx
+        mov     rdi, qword ptr [rbp]
+        mov     qword ptr [rsp + 8], rsi
+        mov     qword ptr [rsp + 16], rdx
         test    rsi, rsi
         je      .LBB4_43
-        test    rbp, rbp
+        test    r15, r15
         je      .LBB4_48
-        mov     rcx, rbp
+        mov     rcx, r15
         call    qword ptr [rip + __rust_realloc@GOTPCREL]
         mov     rcx, rax
         test    rax, rax
         jne     .LBB4_50
-        jmp     .LBB4_65
+        jmp     .LBB4_63
 .LBB4_1:
-        mov     rax, qword ptr [rbx]
-        cmp     r9, 12
+        mov     rsi, qword ptr [rbx]
+        cmp     r12, 12
         jae     .LBB4_2
-        lea     r10, [r9 + 1]
-        test    rax, rax
+        lea     r14, [r12 + 1]
+        test    rsi, rsi
         je      .LBB4_9
-        lea     r14, [r15 + 1]
+        lea     r8, [r13 + 1]
         movabs  rcx, -6148914691236517205
-        mov     rax, r14
+        mov     rax, r8
         mul     rcx
         shr     rdx
         and     rdx, -4
         lea     rax, [rdx + 2*rdx]
-        sub     r14, rax
-        cmp     byte ptr [r13 + r14], 0
-        jne     .LBB4_9
-        mov     byte ptr [r13 + r14], 1
-        add     qword ptr [rbx], -1
-        mov     dword ptr [r12 + 4*r9], 112
+        sub     r8, rax
+        mov     edx, 1
+        mov     ecx, r8d
+        shl     edx, cl
+        bt      ebp, r8d
+        jb      .LBB4_9
```

Comment:
So the main difference appears to be here. Before we had:
```
        cmp     byte ptr [r13 + r14], 0 // if clock_state[next_value]
        jne     .LBB4_9
        mov     byte ptr [r13 + r14], 1 // clock_state[next_value] = true
// call subroutine
        mov     byte ptr [r13 + r14], 0 // clock_state[next_value] = false
```

whereas now we have:
```
        mov     edx, 1 // set 1 in a register so it can be shifted
        mov     ecx, r8d // set clock_state in a register so it can be tested
        shl     edx, cl // 1 << next_value
        bt      ebp, r8d // if clock_state & (1 << next_value) != 0
        jb      .LBB4_9
```

I wonder if it's those `mov`s? 🤷